multiplatform-settings/src/appleMain/kotlin/com/russhwolf/settings/KeychainSettings.kt

```
@@ -0,0 +1,285 @@
+import platform.Security.SecItemAdd
+import platform.Security.SecItemCopyMatching
+import platform.Security.SecItemDelete
+import platform.Security.SecItemUpdate
+import platform.Security.errSecItemNotFound
+import platform.Security.kSecAttrAccount
+import platform.Security.kSecAttrService
+import platform.Security.kSecClass
+import platform.Security.kSecClassGenericPassword
+import platform.Security.kSecMatchLimit
+import platform.Security.kSecMatchLimitAll
+import platform.Security.kSecMatchLimitOne
+import platform.Security.kSecReturnAttributes
+import platform.Security.kSecReturnData
+import platform.Security.kSecValueData
+import platform.darwin.OSStatus
+
+/**
+ * A collection of storage-backed key-value data
+ *
+ * This class allows storage of values with the [Int], [Long], [String], [Float], [Double], or [Boolean] types, using a
+ * [String] reference as a key. Values will be persisted across app launches.
+ *
+ * The specific persistence mechanism is defined using a platform-specific implementation, so certain behavior may vary
+ * across platforms. In general, updates will be reflected immediately in-memory, but will be persisted to disk
+ * asynchronously.
+ *
+ * Operator extensions are defined in order to simplify usage. In addition, property delegates are provided for cleaner
+ * syntax and better type-safety when interacting with values stored in a `Settings` instance.
+ *
+ * The KeychainSettings implementation saves data to the Apple keychain. Data is saved using the generic password type,
+ * where keys are account names and values are treated as passwords. The value passed to the `String` constructor will
+ * be used as the service name. It's also possible to pass custom key-value pairs as attributes that will be added to
+ * every key, if the default behavior does not fit your needs.
+ */
+@ExperimentalSettingsImplementation
+public class KeychainSettings(vararg defaultProperties: Pair<CFStringRef?, CFTypeRef?>) : Settings {
```

Would it be possible to also have a Factory for KeychainSettings? Or is it somewhere and I'm missing it.

---

I've been moving away from the `Factory` interface somewhat as it doesn't really gain you much that you couldn't do by hand with a `() -> Settings` or `(String) -> Settings` lambda. I'll consider adding one here, but you can also roll your own pretty easily if you need it since there's already a constructor that takes a `String` parameter.

---

I like the Factory a lot. My KMP just takes a Factory and then creates instances as needed - I do use about 10-15 instances. On Android + iOS I just pass in the appropriate factory and for tests I can use the MockSettings.Factory.

---

After some consideration, I'm not going to provide a factory here for the same reason there isn't one for Windows: It's not clear what to do when we don't have a name. It's easy to set an application-specific fallback for that case if you create your own factory here, but I'm not sure what a good fallback would be at the library level.

---

Could we throw an exception when the name is null or blank?

---

I'd rather not have a library-provided class that throws there. Since it's easy for you to supply your own, I don't see this as too important an issue.